pulsar-client/src/main/java/org/apache/pulsar/client/impl/MessageCrypto.java

```diff
@@ -0,0 +1,564 @@
+
+import org.apache.pulsar.client.api.CryptoKeyReader;
+import org.apache.pulsar.client.api.PulsarClientException;
+import org.apache.pulsar.client.api.PulsarClientException.CryptoException;
+import org.apache.pulsar.common.api.proto.PulsarApi.KeyByteValue;
+import org.apache.pulsar.common.api.proto.PulsarApi.MessageMetadata;
+import org.apache.pulsar.common.util.collections.ConcurrentOpenHashSet;
+import org.bouncycastle.asn1.ASN1ObjectIdentifier;
+import org.bouncycastle.asn1.pkcs.PrivateKeyInfo;
+import org.bouncycastle.asn1.x509.SubjectPublicKeyInfo;
+import org.bouncycastle.asn1.x9.ECNamedCurveTable;
+import org.bouncycastle.asn1.x9.X9ECParameters;
+import org.bouncycastle.jcajce.provider.asymmetric.ec.BCECPrivateKey;
+import org.bouncycastle.jcajce.provider.asymmetric.ec.BCECPublicKey;
+import org.bouncycastle.jce.provider.BouncyCastleProvider;
+import org.bouncycastle.jce.spec.ECParameterSpec;
+import org.bouncycastle.jce.spec.ECPrivateKeySpec;
+import org.bouncycastle.jce.spec.ECPublicKeySpec;
+import org.bouncycastle.openssl.PEMException;
+import org.bouncycastle.openssl.PEMKeyPair;
+import org.bouncycastle.openssl.PEMParser;
+import org.bouncycastle.openssl.jcajce.JcaPEMKeyConverter;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.protobuf.ByteString;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.PooledByteBufAllocator;
+
+public class MessageCrypto {
+
+    private static final String ECDSA = "ECDSA";
+    private static final String RSA = "RSA";
+    private static final String ECIES = "ECIES";
+    private static final String AESGCM = "AES/GCM/NoPadding";
+
+    private static KeyGenerator keyGenerator;
+    private static final int ivLen = 12;
+    private byte[] iv = new byte[ivLen];
+    private Cipher cipher;
+    MessageDigest digest;
+
+    // Data key which is used to encrypt message
+    private SecretKey dataKey;
+    private LoadingCache<ByteBuffer, SecretKey> dataKeyCache;
+
+    // Map of key name and encrypted gcm key which is sent with encrypted message
+    private ConcurrentHashMap<String, byte[]> encryptedDataKeyMap;
+
+    static final SecureRandom secureRandom;
+    static {
+
+        Security.addProvider(new org.bouncycastle.jce.provider.BouncyCastleProvider());
+        SecureRandom rand = null;
+        try {
+            rand = SecureRandom.getInstance("NativePRNGNonBlocking");
+        } catch (NoSuchAlgorithmException nsa) {
+            rand = new SecureRandom();
+        }
+
+        secureRandom = rand;
+
+        // Initial seed
+        secureRandom.nextBytes(new byte[ivLen]);
+    }
+
+    public MessageCrypto(boolean keyGenNeeded) {
+
+        encryptedDataKeyMap = new ConcurrentHashMap<String, byte[]>();
+        dataKeyCache = CacheBuilder.newBuilder().expireAfterAccess(6, TimeUnit.HOURS)
```

Comment:
if producer regenerates key at every 4 hours and should we also cache up to 4 hours only?

Reply:
We would require few secs/mins more than 4hrs for ideal use case.